### PR TITLE
ASoC: SOF: Intel: hda-codec: Delay the codec device registration

### LINKED ISSUE #4731: [BUG] kernel oops when testing kmod on TGLU_UP_HDA-ipc4
**Describe the bug**
Observed this issue in CI daily test. It happens on TGLU_UP_HDA-IPC4 platform, the reproduction rate is almost 100%. It should be a regression, I will do the bisection next. 

**dmesg**
```
[ 4679.663002] kernel: snd_sof_intel_hda_common:hda_init_caps: sof-audio-pci-intel-tgl 0000:00:1f.3: skipping SoundWire, no links enabled
[ 4679.663139] kernel: snd_sof_intel_hda:hda_codec_probe: sof-audio-pci-intel-tgl 0000:00:1f.3: HDA codec #0 probed OK: response: 10ec0888
[ 4679.668197] kernel: BUG: kernel NULL pointer dereference, address: 0000000000000078
[ 4679.668202] kernel: #PF: supervisor read access in kernel mode
[ 4679.668204] kernel: #PF: error_code(0x0000) - not-present page
[ 4679.668205] kernel: PGD 0 P4D 0 
[ 4679.668207] kernel: Oops: 0000 [#1] PREEMPT SMP NOPTI
[ 4679.668210] kernel: CPU: 3 PID: 61983 Comm: systemd-udevd Not tainted 6.7.0-rc2-g266ee31b71f8 #dev
```

**To Reproduce**
~/sof-test/test-case/check-kmod-load-unload.sh -l 25

**Reproduction Rate**
100%

**Environment**
1) Branch name and commit hash of the 2 repositories: sof (firmware/topology) and linux (kernel driver).
    * Kernel: {[266ee31b71f8](https://github.com/thesofproject/linux/commit/266ee31b71f88434f317d78a9ce3e3aec31fdbd8)}
    * SOF: {[e72056769c8b](https://github.com/thesofproject/sof/commit/e72056769c8bb4b0ee0d2c32a4a0d3a5d5f3181b)}
2) Name of the topology file
    * Topology: {sof-hda-generic-4ch.tplg}
3) Name of the platform(s) on which the bug is observed.
    * Platform: {TGLU_UP_HDA-ipc4}

[dmesg.txt](https://github.com/thesofproject/linux/files/13554036/dmesg.txt)

### PR DESCRIPTION
**Pull request by ujfalusi**
The current code flow is:
1. snd_hdac_device_register()
2. set parameters needed by the hdac driver
3. request_codec_module()
   the hdac driver is probed at this point

During boot the codec drivers are not loaded when the hdac device is
registered, it is going to be probed later when loading the codec module,
which point the parameters are set.

On module remove/insert
rmmod snd_sof_pci_intel_tgl
modprobe snd_sof_pci_intel_tgl

The codec module remains loaded and the driver will be probed when the
hdac device is created right away, before the parameters for the driver
has been configured:

1. snd_hdac_device_register()
   the hdac driver is probed at this point
2. set parameters needed by the hdac driver
3. request_codec_module()
   will be a NOP as the module is already loaded

Move the snd_hdac_device_register() later, to be done right before
requesting the codec module to make sure that the parameters are all set
before the device is created:

1. set parameters needed by the hdac driver
2. snd_hdac_device_register()
3. request_codec_module()

This way at the hdac driver probe all parameters will be set in all cases.

Link: https://github.com/thesofproject/linux/issues/4731
Fixes: a0575b4add21 ("ASoC: hdac_hda: Conditionally register dais for HDMI and Analog")

### CODE REVIEW

File: sound/soc/sof/intel/hda-codec.c [hunk 1]
```diff
@@ -54,8 +54,16 @@ static int request_codec_module(struct hda_codec *codec)
...
 {
-	int ret = request_codec_module(codec);
+	int ret;
+
+	ret = snd_hdac_device_register(&codec->core);
+	if (ret) {
+		dev_err(&codec->core.dev, "failed to register hdac device\n");
+		put_device(&codec->core.dev);
+		return ret;
+	}
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-codec.c:64]: it's hard to follow the commit message 

The current code flow is:
snd_hdac_device_register()
set parameters needed by the hdac driver <<<< I see nothing between register() and request()?
request_codec_module()


Move the snd_hdac_device_register() later, to be done right before
requesting the codec module to make sure that the parameters are all set
before the device is created. <<< again it's not clear where those parameters are set??

**Comment by ujfalusi** [comment]
Changes since v1:
- reword the commit message

**Review by bardliao — APPROVED** [review]

**Review by plbossart — APPROVED** [review]
